File: src/ReetCode/Easy/101to200/Q104/Q104_Maximum Depth of Binary Tree.py
"""
# Author    : Lee In Bok 
# Date      : 2023.12.01(Fri)
# Spend Time: 04m 26s
# Runtime   : 35 ms (Beats 97.39%)
# Memory    : 18.9 MB (Beats 6.48%)
# Algoritm  : Tree
"""

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def maxDepth(self, root: Optional[TreeNode]) -> int:
        max_depth = 0

        if not root:
            return 0

        def order(depth, root):
            nonlocal max_depth
            max_depth = max(depth, max_depth)
            if root:
                if root.left:
                    order(depth + 1, root.left)
                if root.right:
                    order(depth + 1, root.right)
        
        order(1, root)
        return max_depth
        

File: CodeTest/src/ReetCode/Easy/101to200/Q104/Q104_MaximumDepthOfBinaryTree.java
/**
* Author    : Lee In Bok 
* Date      : 2023.12.01(Fri)
* Spend Time: 04m 26s
* Runtime   : 0 ms (Beats 100%)
* Memory    : 41.9 MB (Beats 23.92%)
* Algoritm  : Tree
 */

// class Solution {
//     int max = 0;

//     public int maxDepth(TreeNode root) {
//         if(root == null){
//             return 0;
//         }
        
//         order(1, root);

//         return max;
//     }

//     public void order(int depth, TreeNode root){
//         max = Math.max(depth, max);

//         if(root != null){
//             if(root.left != null){
//                 order(depth + 1, root.left);
//             }

//             if(root.right != null){
//                 order(depth + 1, root.right);
//             }
//         }
//     }
// }